sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,752 @@
+  az_json_builder json_builder;
+  if (az_failed(rc = build_reported_property(&json_builder, desired_temp)))
+  {
+    return rc;
+  }
+  az_span json_payload = az_json_builder_get_json(&json_builder);
+
+  printf("Payload: %.*s\n", az_span_size(json_payload), (char*)az_span_ptr(json_payload));
+
+  // Publish the reported property payload to IoT Hub
+  if ((rc = MQTTClient_publish(
+           mqtt_client,
+           reported_property_topic,
+           az_span_size(json_payload),
+           az_span_ptr(json_payload),
+           0,
+           0,
+           NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to publish reported property, return code %d\n", rc);
+    return rc;
+  }
+  return rc;
+}
+
+// Parse the desired temperature property from the incoming JSON
+static az_result parse_desired_temperature_property(az_span twin_span, double* parsed_value)
+{
+  az_json_parser jp;
+  az_json_token_member tm;
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, twin_span));
+  AZ_RETURN_IF_FAILED(az_json_parser_parse_token(&jp, &tm.token));
+  if (tm.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  while (az_succeeded(az_json_parser_parse_token_member(&jp, &tm)))
+  {
+    if (az_span_is_content_equal(desired_temp_property_name, tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_number(&tm.token, parsed_value));
+    }
+
+    // else ignore token.
+  }
+
+  return AZ_OK;
+}
+
+// Callback for incoming MQTT messages
+static int on_received(void* context, char* topicName, int topicLen, MQTTClient_message* message)
+{
+  (void)context;
+
+  if (topicLen == 0)
+  {
+    // The length of the topic if there are one or more NULL characters embedded in topicName,
+    // otherwise topicLen is 0.
+    topicLen = (int)strlen(topicName);
+  }
+
+  printf("Topic: %s\n", topicName);
+
+  az_span topic_span = az_span_init((uint8_t*)topicName, topicLen);
+
+  // Parse the incoming message topic and check which feature it is for
+  az_iot_hub_client_twin_response twin_response;
+  az_iot_hub_client_method_request method_request;
+
+  if (az_succeeded(
+          az_iot_hub_client_twin_parse_received_topic(&client, topic_span, &twin_response)))
+  {
+    printf("Twin Message Arrived\n");
+
+    // Determine what type of incoming twin message this is. Print relevant data for the message.
+    switch (twin_response.response_type)
+    {
+      // A response from a twin GET publish message with the twin document as a payload.
+      case AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_GET:
+        printf("A twin GET response was received\n");
+        if (message->payloadlen)
+        {
+          printf("Payload:\n%.*s\n", message->payloadlen, (char*)message->payload);
+        }
+        break;
+      // An update to the desired properties with the properties as a JSON payload.
+      case AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES:
```

Comment:
Will `AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES` be triggered always or only if the server did a twin update *while* the device was connected?

If it's only when the device was connected, then I think we need equivalent parsing code up in `AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_GET`?  The scenario I'm looking for is device connects and it gets its desired properties, even if server never modifies them during the lifetime of the given connection, but we'd still want to set the tempearute to be whatever server had sent previously.

Reply:
anything that comes into the device on the topic
`$iothub/twin/PATCH/properties/desired/#`
will get labeled with `AZ_IOT_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES`. My understanding is that is only when the service changes a desired property and it sends that message.

Reply:
OK in that case then I think you need to get the twin explicitly and then parse out its "desired" since that state may have been changed when device was offline.  The C SDK does GetTwin() implicitly on the callback for getting twin notifications; embedded wisely doesn't follow that pattern but I think it needs something similar so-implemented here.

Reply:
Just FYI.
@danewalton  I have the changes in convenience layer to parse full twin document and partial document. 
https://expresslogic.visualstudio.com/X-Ware/_git/azure-services-port?path=%2Fsdk%2Fsamples%2Fsample_azure_iot_embedded_sdk_pnp.c&version=GBdevelop_higupt_no_component&line=259&lineEnd=260&lineStartColumn=1&lineEndColumn=1&lineStyle=plain